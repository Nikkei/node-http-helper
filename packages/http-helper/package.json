{
    "name": "@nikkei/http-helper",
    "version": "2.4.1",
    "type": "module",
    "files": [
        "__dist/",
        "LICENSE",
        "package.json",
        "README.md"
    ],
    "typings": "./__dist/mod.d.ts",
    "exports": {
        ".": {
            "types": "./__dist/mod.d.ts",
            "default": "./__dist/mod.js"
        },
        "./auth": {
            "types": "./__dist/auth/mod.d.ts",
            "default": "./__dist/auth/mod.js"
        },
        "./cache": {
            "types": "./__dist/cache/mod.d.ts",
            "default": "./__dist/cache/mod.js"
        },
        "./csp": {
            "types": "./__dist/csp/mod.d.ts",
            "default": "./__dist/csp/mod.js"
        },
        "./sec_fetch": {
            "types": "./__dist/sec_fetch/mod.d.ts",
            "default": "./__dist/sec_fetch/mod.js"
        }
    },
    "scripts": {
        "clean": "npm run clean:dist && npm run clean:tsbuildinfo",
        "clean:dist": "rm -rf __dist/",
        "clean:tsbuildinfo": "rm -rf tsconfig.tsbuildinfo",
        "prebuild": "npm run clean",
        "build": "tsc --project ."
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/Nikkei/node-http-helper.git"
    },
    "author": "Nikkei Inc.",
    "license": "MIT",
    "bugs": {
        "url": "https://github.com/Nikkei/node-http-helper/issues"
    },
    "homepage": "https://github.com/Nikkei/node-http-helper",
    "dependencies": {
        "option-t": "^39.0.2"
    }
}
