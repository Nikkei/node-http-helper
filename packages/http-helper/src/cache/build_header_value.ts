/**
 *  @see
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-cache-control
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-collected-abnf
 *
 *  @throws {RangeError}
 *      This throws if the caller calls without any arguments.
 *      Then the caller does not have to use this.
 */
export function buildCacheControlValue(...items: Array<string>): string {
    if (items.length <= 0) {
        throw new RangeError(`items must be > 0`);
    }

    const result = items.join(', ');
    return result;
}
