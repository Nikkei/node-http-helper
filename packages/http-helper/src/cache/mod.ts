//  @see
//  - [RFC 9111 HTTP Caching](https://www.rfc-editor.org/rfc/rfc9111.html)
//  - [RFC 9213 Targeted HTTP Cache Control](https://www.rfc-editor.org/rfc/rfc9213)

export * as RequestDirective from './request_directive.js';
