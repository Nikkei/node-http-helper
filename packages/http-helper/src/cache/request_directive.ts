//  @see
//  - https://www.rfc-editor.org/rfc/rfc9111.html#name-request-directives
//  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#request_directives
import {
    TOKEN_MAX_AGE,
    TOKEN_MAX_STALE,
    TOKEN_MIN_FRESH,
    TOKEN_NO_CACHE,
    TOKEN_NO_STORE,
    TOKEN_NO_TRANSFORM,
    TOKEN_ONLY_IF_CACHED,
    TOKEN_STALE_IF_ERROR,
} from './internal/token_intern.js';

/**
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-max-age
 */
export const MAX_AGE = TOKEN_MAX_AGE;

/**
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-max-stale
 */
export const MAX_STALE = TOKEN_MAX_STALE;

/**
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-min-fresh
 */
export const MIN_FRESH = TOKEN_MIN_FRESH;

/**
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-no-cache
 */
export const NO_CACHE = TOKEN_NO_CACHE;

/**
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-no-store
 */
export const NO_STORE = TOKEN_NO_STORE;

/**
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-no-transform
 */
export const NO_TRANSFORM = TOKEN_NO_TRANSFORM;

/**
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-only-if-cached
 */
export const ONLY_IF_CACHED = TOKEN_ONLY_IF_CACHED;

/**
 *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#stale-if-error
 *  - https://datatracker.ietf.org/doc/rfc5861/
 */
export const STALE_IF_ERROR = TOKEN_STALE_IF_ERROR;
