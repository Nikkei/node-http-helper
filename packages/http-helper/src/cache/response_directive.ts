//  @see
//  - https://www.rfc-editor.org/rfc/rfc9111.html#name-response-directives
//  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#response_directives
import {
    TOKEN_IMMUTABLE,
    TOKEN_MAX_AGE,
    TOKEN_MUST_REVALIDATE,
    TOKEN_MUST_UNDERSTAND,
    TOKEN_NO_CACHE,
    TOKEN_NO_STORE,
    TOKEN_NO_TRANSFORM,
    TOKEN_PRIVATE,
    TOKEN_PROXY_REVALIDATE,
    TOKEN_PUBLIC,
    TOKEN_STALE_IF_ERROR,
    TOKEN_STALE_WHILE_REVALIDATE,
    TOKEN_S_MAXAGE,
} from './internal/token_intern.js';

/**
 *  - https://datatracker.ietf.org/doc/rfc8246/
 *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#immutable
 */
export const IMMUTABLE = TOKEN_IMMUTABLE;

/**
 *  https://www.rfc-editor.org/rfc/rfc9111.html#name-max-age-2
 */
export const MAX_AGE = TOKEN_MAX_AGE;

/**
 *  https://www.rfc-editor.org/rfc/rfc9111.html#name-must-revalidate
 */
export const MUST_REVALIDATE = TOKEN_MUST_REVALIDATE;

/**
 *  https://www.rfc-editor.org/rfc/rfc9111.html#name-must-understand
 */
export const MUST_UNDERSTAND = TOKEN_MUST_UNDERSTAND;

/**
 *  https://www.rfc-editor.org/rfc/rfc9111.html#name-no-cache-2
 */
export const NO_CACHE = TOKEN_NO_CACHE;

/**
 *  https://www.rfc-editor.org/rfc/rfc9111.html#name-no-store-2
 */
export const NO_STORE = TOKEN_NO_STORE;

/**
 * https://www.rfc-editor.org/rfc/rfc9111.html#name-no-transform-2
 */
export const NO_TRANSFORM = TOKEN_NO_TRANSFORM;

/**
 *  https://www.rfc-editor.org/rfc/rfc9111.html#name-private
 */
export const PRIVATE = TOKEN_PRIVATE;

/**
 *  https://www.rfc-editor.org/rfc/rfc9111.html#name-proxy-revalidate
 */
export const PROXY_REVALIDATE = TOKEN_PROXY_REVALIDATE;

/**
 *  https://www.rfc-editor.org/rfc/rfc9111.html#name-public
 */
export const PUBLIC = TOKEN_PUBLIC;

/**
 *  https://www.rfc-editor.org/rfc/rfc9111.html#name-s-maxage
 */
export const S_MAXAGE = TOKEN_S_MAXAGE;

/**
 *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#stale-if-error
 *  - https://datatracker.ietf.org/doc/rfc5861/
 */
export const STALE_IF_ERROR = TOKEN_STALE_IF_ERROR;

/**
 *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#stale-while-revalidate
 *  - https://datatracker.ietf.org/doc/rfc5861/
 */
export const STALE_WHILE_REVALIDATE = TOKEN_STALE_WHILE_REVALIDATE;
