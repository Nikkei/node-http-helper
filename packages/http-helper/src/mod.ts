export * as Auth from './auth/mod.js';
export * as Cache from './cache/mod.js';
export * as ClearSiteData from './clear_site_data.js';
export * as CSP from './csp/mod.js';
export * as HttpHeaderName from './http_header_name.js';
export * as HttpMethod from './http_method.js';
export * as HttpStatus from './http_status_code.js';
export * as Mime from './mime.js';
export { SecFetchDest, SecFetchMode, SecFetchSite, SecFetchUser } from './sec_fetch/mod.js';
