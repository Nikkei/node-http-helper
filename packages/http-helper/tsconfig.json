{
    "extends": "./tsconfig.base.json",
    "compilerOptions": {
        "outDir": "./__obj"
    },
    "include": [
        // @prettier-ignore
        "src/**/*",
        "__tests__/**/*"
    ]
}
