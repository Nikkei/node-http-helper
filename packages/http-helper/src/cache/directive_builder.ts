import { assertIsInteger, assertIsNotNegative } from './internal/check_digit.js';
import { MAX_STALE, MIN_FRESH } from './request_directive.js';
import { MAX_AGE, STALE_IF_ERROR, STALE_WHILE_REVALIDATE, S_MAXAGE } from './response_directive.js';

/**
 *  @see
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-max-age
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-max-age-2
 *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#max-age
 *
 *  @throws {TypeError}
 *      The argument is not safe integer. It's not spec compatible.
 *  @throws {RangeError}
 *      The argument is < 0. It's not spec compatible.
 */
export function maxAge(seconds: number) {
    // See
    //  - https://www.rfc-editor.org/rfc/rfc9111#section-5.2.1.1-1
    //  - https://www.rfc-editor.org/rfc/rfc9111#section-1.2.2
    assertIsNotNegative(seconds);
    assertIsInteger(seconds);

    return buildKeyValue(MAX_AGE, seconds);
}

/**
 *  @see
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-max-stale
 *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#max-stale
 *
 *  @throws {TypeError}
 *      The argument is not safe integer. It's not spec compatible.
 *  @throws {RangeError}
 *      The argument is < 0. It's not spec compatible.
 */
export function maxStale(seconds: number) {
    // See
    //  - https://www.rfc-editor.org/rfc/rfc9111#section-5.2.1.1-1
    //  - https://www.rfc-editor.org/rfc/rfc9111#section-1.2.2
    assertIsNotNegative(seconds);
    assertIsInteger(seconds);

    return buildKeyValue(MAX_STALE, seconds);
}

/**
 *  @see
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-min-fresh
 *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#min-fresh
 *
 *  @throws {TypeError}
 *      The argument is not safe integer. It's not spec compatible.
 *  @throws {RangeError}
 *      The argument is < 0. It's not spec compatible.
 */
export function minFresh(seconds: number) {
    // See
    //  - https://www.rfc-editor.org/rfc/rfc9111#section-5.2.1.1-1
    //  - https://www.rfc-editor.org/rfc/rfc9111#section-1.2.2
    assertIsNotNegative(seconds);
    assertIsInteger(seconds);

    return buildKeyValue(MIN_FRESH, seconds);
}

/**
 *  @see
 *  - https://www.rfc-editor.org/rfc/rfc9111.html#name-s-maxage
 *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#s-maxage
 *
 *  @throws {TypeError}
 *      The argument is not safe integer. It's not spec compatible.
 *  @throws {RangeError}
 *      The argument is < 0. It's not spec compatible.
 */
export function sMaxAge(seconds: number) {
    // See
    //  - https://www.rfc-editor.org/rfc/rfc9111#section-5.2.1.1-1
    //  - https://www.rfc-editor.org/rfc/rfc9111#section-1.2.2
    assertIsNotNegative(seconds);
    assertIsInteger(seconds);

    return buildKeyValue(S_MAXAGE, seconds);
}

/**
 *  @see
 *  - https://datatracker.ietf.org/doc/html/rfc5861#page-3
 *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#stale-if-error
 *
 *  @throws {TypeError}
 *      The argument is not safe integer. It's not spec compatible.
 *  @throws {RangeError}
 *      The argument is < 0. It's not spec compatible.
 */
export function staleIfError(seconds: number) {
    // See
    // https://datatracker.ietf.org/doc/html/rfc2616#section-3.3.2
    // https://datatracker.ietf.org/doc/html/rfc5861#section-4.1
    assertIsNotNegative(seconds);
    assertIsInteger(seconds);

    return buildKeyValue(STALE_IF_ERROR, seconds);
}

/**
 *  @see
 *  - https://datatracker.ietf.org/doc/html/rfc5861#section-3
 *  - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#stale-while-revalidate
 *
 *  @throws {TypeError}
 *      The argument is not safe integer. It's not spec compatible.
 *  @throws {RangeError}
 *      The argument is < 0. It's not spec compatible.
 */
export function staleWhileRevalidate(seconds: number) {
    // See
    // https://datatracker.ietf.org/doc/html/rfc2616#section-3.3.2
    // https://datatracker.ietf.org/doc/html/rfc5861#section-3.1
    assertIsNotNegative(seconds);
    assertIsInteger(seconds);

    return buildKeyValue(STALE_WHILE_REVALIDATE, seconds);
}

function buildKeyValue<const TDirName extends string>(key: TDirName, numericValue: number): `${TDirName}=${string}` {
    return `${key}=${String(numericValue)}`;
}
