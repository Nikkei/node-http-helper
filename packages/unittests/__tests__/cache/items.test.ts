import { test, expect } from 'vitest';

import { RequestDirective, ResponseDirective, Value } from '@nikkei/http-helper/cache';

test('RequestDirective entries', () => {
    expect(RequestDirective).toMatchSnapshot();
});

test('ResponseDirective entries', () => {
    expect(ResponseDirective).toMatchSnapshot();
});

test('Value entries', () => {
    expect(Value).toMatchSnapshot();
});
