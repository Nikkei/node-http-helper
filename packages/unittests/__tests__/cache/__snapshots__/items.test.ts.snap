// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RequestDirective entries 1`] = `
{
  "MAX_AGE": "max-age",
  "MAX_STALE": "max-stale",
  "MIN_FRESH": "min-fresh",
  "NO_CACHE": "no-cache",
  "NO_STORE": "no-store",
  "NO_TRANSFORM": "no-transform",
  "ONLY_IF_CACHED": "only-if-cached",
  "STALE_IF_ERROR": "stale-if-error",
}
`;

exports[`ResponseDirective entries 1`] = `
{
  "IMMUTABLE": "immutable",
  "MAX_AGE": "max-age",
  "MUST_REVALIDATE": "must-revalidate",
  "MUST_UNDERSTAND": "must-understand",
  "NO_CACHE": "no-cache",
  "NO_STORE": "no-store",
  "NO_TRANSFORM": "no-transform",
  "PRIVATE": "private",
  "PROXY_REVALIDATE": "proxy-revalidate",
  "PUBLIC": "public",
  "STALE_IF_ERROR": "stale-if-error",
  "STALE_WHILE_REVALIDATE": "stale-while-revalidate",
  "S_MAXAGE": "s-maxage",
}
`;

exports[`Value entries 1`] = `
{
  "maxAge": [Function],
  "maxStale": [Function],
  "minFresh": [Function],
  "sMaxAge": [Function],
  "staleIfError": [Function],
  "staleWhileRevalidate": [Function],
}
`;
